OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamageState.cs

```diff
@@ -32,41 +32,53 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
 		[Desc("Are upgrades irrevocable once the conditions have been met?")]
 		public readonly bool GrantPermanently = false;
 
-		public object Create(ActorInitializer init) { return new UpgradeOnDamage(init.Self, this); }
+		public object Create(ActorInitializer init) { return new UpgradeOnDamageState(init.Self, this); }
 	}
 
-	public class UpgradeOnDamage : INotifyDamageStateChanged
+	public class UpgradeOnDamageState : INotifyDamageStateChanged, INotifyCreated
 	{
-		readonly UpgradeOnDamageInfo info;
+		readonly UpgradeOnDamageStateInfo info;
 		readonly UpgradeManager um;
+		readonly Health health;
 		bool granted;
 
-		public UpgradeOnDamage(Actor self, UpgradeOnDamageInfo info)
+		public UpgradeOnDamageState(Actor self, UpgradeOnDamageStateInfo info)
 		{
 			this.info = info;
-			um = self.TraitOrDefault<UpgradeManager>();
+			um = self.Trait<UpgradeManager>();
+			health = self.Trait<Health>();
 		}
 
-		void INotifyDamageStateChanged.DamageStateChanged(Actor self, AttackInfo e)
+		void INotifyCreated.Created(Actor self)
+		{
+			if (!granted)
```

Comment:
Isn't this check redundant? I don't see a way how `granted` could become true before `Created` runs.